xds/internal/balancer/ringhash/ring.go

```diff
@@ -92,16 +92,19 @@ func newRing(subConns *resolver.AddressMap, minRingSize, maxRingSize uint64, log
 	//
 	// A hash is generated for each item, and later the results will be sorted
 	// based on the hash.
-	var (
-		idx       int
-		targetIdx float64
-	)
+	var currentHashes, targetHashes float64
 	for _, scw := range normalizedWeights {
-		targetIdx += scale * scw.weight
-		for float64(idx) < targetIdx {
-			h := xxhash.Sum64String(scw.sc.addr + strconv.Itoa(idx))
-			items = append(items, &ringEntry{idx: idx, hash: h, sc: scw.sc})
+		targetHashes += scale * scw.weight
+		// This index ensures that ring entries corresponding to the same
+		// address hash to different values. And since this index is
+		// per-address, these entries hash to the same value across address
+		// updates.
+		idx := 0
+		for currentHashes < targetHashes {
```

Comment:
Since `scw.weight` is not a whole number, `targetHashes` is not a whole number either. In order to ensure that we end up with the correct number of items in the ring, we need to keep a running sum.

I based this change on C-core and Java's implementation. Looks like envoy also does the same: https://github.com/envoyproxy/envoy/blob/main/source/common/upstream/ring_hash_lb.cc#L159.